### ISSUE TITLE
eBPF support for TCP payload append feature.

### ISSUE THREAD

**Issue by dhawaskar** [cilium/ebpf#339]
Hi everyone,
I wanted to bring in new bug or future that could be added in eBPF. This future would help to support INT in programmable switches.

In the kernel by default TCP uses CHECKSUM_PARTIAL flag to compute the outer pseudo csum, after the TC-eBPF, the csum is updated in the driver using  skb->csum_start and skb->csum_offset field. 

Now if we want to append to the existing TCP payload, only way in eBPF  at TC is using bpf_skb_adjust_room API that allows appending only after ipv4 header (There is no API to adjust room after L4 header). Hence only way to append after L4 header is to allocate space after L3 and pull the TCP header up and append the data in hole created. But this will modify the TCP  skbuffer fileds skb->csu_start and skb->csum_offset that are not available in TC-eBPF . This issue has another issue within itself that payload not accessible for TCP in 5.x kernel ,only in 4.x kernel. Even if we figure out a way to compute checksum in eBPF,below netdirver layer modifies appended data as skb->csu_start and skb->csum_offset fields and corrupt csum written in TC layer. 

Another work around this issue is to force kernel to use CHECKSUM_UNNECESSARY flag. Using this kernel modification we have complete csum for TCP, hence we can append data and modify the TCP csum. But again since we adding the added length in TC and kernel is not aware of it, there create re-transmission and packet loss problem. Another major issue with this workaround is the TCP throughput dropped by 80% using iperf3.

Hence to do such modification for TCP telemetry in eBPF with software switches, we need to add new eBPF support as kernel workaround can create csum issues for TCP.

If anyone has idea or though of the problem. Please flash some light. I will also create a new bug for this. I have also posted details of beginning of problem with code snippet above.

Sandesh

**Comment by lmb** [comment]
Hi Sandesh: this repository is used to develop a user space library that gives you access to the BPF facilities of the Linux kernel. You should direct your proposal towards the BPF mailing list, see https://www.kernel.org/doc/html/latest/bpf/bpf_devel_QA.html

Please stop creating issues in this project to discuss issues with upstream BPF.

**Comment by stv0g** [comment]
Hi @dhawaskar,

I am facing exactly the same issue. Have you already found a solution for your problem?

**Comment by alepvn33** [comment]
Hello @dhawaskar and @stv0g, I have the same problem. Have you found a solution?
